### PR TITLE
ENH small cleaning and optimizations accross the repo 

### PR DESCRIPTION
**Pull request by fcharras**
- [x] Replace signature exposing sizes with a signature exposing one argument `shape`
- [x] Where applicable, use 2D grid of work items rather than 1D + divisions 
    + note: found this piece of information: https://stackoverflow.com/a/15044884 suggesting that:
        + 2D grid really are better for performance because doing // and % ops in the kernel is expensive
        + it indexes work items in a "row major order" meaning that items in the same row belong to the same sub group (/warp)
    + even if the previous information applies to cuda. I assume it's also true for numba-dpex/ sycl
- [x] some kernels are now written only for a given shape and reused for other dimension with reshaping tricks.
- [x] also use 2D grids for kmeans kernels
- [x] document better the use of `centroids_private_copies_max_cache_occupancy` and improve the heuristic using `device.max_compute_units`
- [x] a couple more nitpicks

~~It's WIP because using 2D groups for the sum over axis 0 kernel seems to trigger weirg bugs like in https://github.com/IntelPython/numba-dpex/issues/892 in some of the tests.~~

Out of WIP, remaining issues were unrelated, PR is green.

edit: confirmed affected by https://github.com/IntelPython/numba-dpex/issues/906

### CODE REVIEW

**Comment by fcharras** [comment]
Out of WIP. See the top level post for the list of changes.

File: sklearn_numba_dpex/kmeans/drivers.py
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:473]: I do not find this chaining super intuitive. Can you extract it above just above the call to this function and give it a name? :slightly_smiling_face:

File: sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py:239]: Can you give this a name?

File: sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py:111]: Identically, is it possible to give this a name?

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/compute_labels.py:175]: Identically, is it possible to give this a name?

File: sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py:364]: Is it possible to avoid using a modulo, here?

**Review by jjerphan — COMMENTED** [review]
Thank you for this PR of general improvements, @fcharras. :slightly_smiling_face: 

Here is a first pass.

File: sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py:364]: I don't think so :-(

File: sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py]: I agree but I don't really now how to it clear, it's twisted that the shape of the grid is adapted to the sliding window on centroids, but each work item apply to one single input sample.

File: sklearn_numba_dpex/common/kernels.py
Comment (fcharras) [inline sklearn_numba_dpex/common/kernels.py:753]: There's probably some bug in numba_dpex or with the cpu backend here that causes branching to not be correctly executed in very specific cases, for this very particular condition, on cpu. I had no luck when trying to have a minimal reproducer and I can't see failures on gpu. So I'd be in favor of letting this pass through in the meantime. Note that this also fixes the case with `work_group_size=1` that I added back to the unit tests.

Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:753]: Well it's still wrong on the CI. :shrug:

**Comment by fcharras** [comment]
Something is wrong with `sum(axis=1)` on CPU when `work_group_size` is `>=8`. Maybe another bug in the JIT or something we don't understand about group sizes on CPU (or a bit of both). In both cases the investigation looks complicated and CPU is not the priority target so the latest commit propose forcing `work_group_size == 8` with `work_group_size` was set to `max`.

I'll also run this branch on the dev cloud with the flex170 to see if it shows any performance improvement, and play with the group size to see if it affects performance more than it seems to do on local iGPU.

File: sklearn_numba_dpex/common/kernels.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/kernels.py:594]: Nit: drop the leading `_` in the variable name if you don't expect a name conflict.

File: sklearn_numba_dpex/common/kernels.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/kernels.py:633]: Same comment here.

File: sklearn_numba_dpex/common/kernels.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/kernels.py:717]: black:

```suggestion
        )
```

which makes me think that I wanted to open a PR to fix the black problem but completely forgot about it. I will wait for this PR to me finalized first though.

File: sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py:238]: Are you sure this change is correct? Don't we need a bound check on the value of `local_col_idx`? I would have maybe expected:

```suggestion
            if local_row_idx == zero_as_a_long and local_col_idx < window_n_centroids:
```

assuming that `window_of_centroids_half_l2_norms` is raveled window of logical shape `(1, window_n_centroids)`, but the comment should probably be updated to better explain the interaction between `local_row_idx`, `local_col_idx` and the shape of the window array.

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/compute_labels.py:93]: I admit I am lost here. Could you add some comments to explain how `sub_group_size` and the 2d local indexing logic work?

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/compute_labels.py:47]:
```suggestion
    work_group_shape = (centroids_window_height, centroids_window_width)
```

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/compute_labels.py:33]: Maybe we could do:

```suggestion
    centroids_window_width = sub_group_size
```

and completely drop the `window_n_centroids` variable to remove on indirection?

File: sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py:123]:
```suggestion
    # refer too, please read the inline comments in the kernel definition.
```

to avoid duplicating "more details".

File: sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py:133]:
```suggestion
    # copy, and several such sets can be assigned to the same private copy. Thus, at
```

File: sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py:229]: Making the shapes explicit after IN / OUT would probably help.

File: sklearn_numba_dpex/kmeans/kernels/utils.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/utils.py:232]: Isn't there an indentation problem here?

File: sklearn_numba_dpex/kmeans/kernels/utils.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/utils.py:234]: This seems redundant with the `if item_idx >= n_sums:` early return at the beginning of the kernel.

File: sklearn_numba_dpex/kmeans/kernels/utils.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels/utils.py:217]: Maybe rename to `centroids_t_flattened`

**Review by ogrisel — COMMENTED** [review]
Overall, the changes improve the readability a lot (I think). Looking forward to the performance impacts (once the cause of the bug has been found).

I have seen a bunch of fishing things (see some of the comments below). Maybe this can explain the broken tests?

**Comment by fcharras** [comment]
I've reduced the last failure on the pipeline to what appear to be a JIT issue. The merge will be blocked until it's fixed. It's likely the same JIT issue we've seen before and it gave a much nicer reproducer. See https://github.com/IntelPython/numba-dpex/issues/906 .

File: sklearn_numba_dpex/common/kernels.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:296]: Should the message be changed here?

File: sklearn_numba_dpex/common/kernels.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:415]: The last comment likely is off due to black formatting.

What do you think of this suggestion?
```suggestion
    kernels = (partial_sum_reduction, partial_sum_reduction_nofunc)
    shape_update_fn = (get_result_shape, get_global_size)
    return (
        work_group_shape,
        (
            kernels,
            reduction_block_size,
        ),
        shape_update_fn,
    )
```

File: sklearn_numba_dpex/common/kernels.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:629]: The previous suggestion also applies here.

File: sklearn_numba_dpex/common/kernels.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:751]: Can you add inline comment to give some precisions regarding this branch?

**Review by jjerphan — APPROVED** [review]
LGTM once @ogrisel's comment have been addressed.

File: sklearn_numba_dpex/common/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:92]: The intent here was to mean that the broadcasted data is `1d`. Would `1d_to_2d` make it clearer ?

File: sklearn_numba_dpex/common/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:594]: I'll revert this `_is_cpu` change, in fact it does not fix the JIT issue. We must wait for the root cause to be addressed in `numba_dpex`

File: sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py:238]: Well spotted, I think your suggestion is right. There is a risk to read `centroids_half_l2_norm` array in global memory outside of its boundaries. Not sure why execution on cpu doesn't segfaults :thinking: 

My guess is that the compiler detected that the corresponding write in local memory is not used later on (which is True) and optimized in a way that skipped the risk of segfault.

Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py:238]: Will update comment accordingly

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/compute_labels.py:93]: Will add a dosctring in the `lloyd` kernel (where most of the explanations for similar kernels are located)

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/compute_labels.py:47]: or let's just drop `centroids_window_width` variable ? (that was my choice for other files, I probably forgot this one)

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/compute_labels.py:33]: I propose removing `centroids_window_width` completely here. (see previous comment)

Now, regarding using `sub_group_size` or `window_n_centroids` I don't know what is better. In the current iteration I've favored using `window_n_centroids`.

File: sklearn_numba_dpex/kmeans/kernels/utils.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:232]: good catch :relieved: TY

File: sklearn_numba_dpex/common/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:296]: In my mind `work_group_size=math.prod(work_group_shape)` so I think it's good.

There is this nuance because I have introduced the syntax `work_group_size=max` where we try to automatically use the maximum possible value for `work_group_size`. This change enable checking that `max` does not result in `work_group_size=1`.

File: sklearn_numba_dpex/common/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:751]: I'll revert this `_is_cpu` change, in fact it does not fix the JIT issue. We must wait for the root cause to be addressed in `numba_dpex`

**Comment by fcharras** [comment]
TY for the careful review @ogrisel @jjerphan the last commit should address your suggestions and I've answered some of your comments.

**Comment by jjerphan** [comment]
Perfect. :+1: 

I leave @ogrisel merge.

**Comment by ogrisel** [comment]
The tests are still red. Ok for merge once fixed.

**Comment by fcharras** [comment]
The tests that fail because of https://github.com/IntelPython/numba-dpex/issues/906 we can merge as soon as it's fixed and we can bump

If there's a conflict with other branches before that we can merge except the diff on the 2d sum kernel.

**Comment by fcharras** [comment]
Closed via https://github.com/soda-inria/sklearn-numba-dpex/pull/98 and https://github.com/soda-inria/sklearn-numba-dpex/pull/96

**Pull request closed without merging**
